### PR TITLE
Fix stdlib exception leak in `bytes::sub`

### LINKED ISSUE #1436: `bytes::sub` can throw non-HILTI exception
The following directly throws a `std::out_of_range` while it should throw a HILTI exception:

```ruby
# file: foo.spicy
module foo;

b"".sub(10, 11);
```

```console
$ spicyc -dj foo.spicy
[error] uncaught exception of type std::out_of_range: basic_string
```

### PR DESCRIPTION
**Pull request by bbannier**
Closes #1436.

### CODE REVIEW

File: hilti/runtime/include/types/bytes.h [hunk 1]
```diff
@@ -294,7 +295,13 @@ class Bytes : protected std::string {
      * @param offset of one byeond end of subrage
      * @return a `Bytes` instance for the subrange
      */
-    Bytes sub(Offset from, Offset to) const { return {substr(from, to - from)}; }
+    Bytes sub(Offset from, Offset to) const {
+        try {
+            return {substr(from, to - from)};
+        } catch ( const std::out_of_range& ) {
+            throw InvalidArgument(fmt("start index %s out of range for bytes with length %d", from, size()));
```
Comment (rsmmr) [inline hilti/runtime/include/types/bytes.h:302]: We have our own `OutOfRange`, I'd throw that.

**Review by rsmmr — APPROVED** [review]
I don't quite see why we the 1st commit is needed to fix this? (But fine either way.)

**Comment by bbannier** [comment]
> I don't quite see why we the 1st commit is needed to fix this? (But fine either way.)

The idea was to minimize places where we would have called functions of `std::string` and would need a catch block. With that patch we now only call `substr` exactly once and handle the exception there.

**Comment by bbannier** [comment]
Also backported to `release/1.5`.

**Pull request merged**